### PR TITLE
Split overloaded transfer / timer task APIs

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
Split overloaded transfer / timer task APIs
* Split TransferTasksFilter into TransferTasksFilter and TransferTasksRangeFilter
* Split SelectFromTransferTasks into SelectFromTransferTasks and RangeSelectFromTransferTasks
* Split DeleteFromTransferTasks into DeleteFromTransferTasks and RangeDeleteFromTransferTasks
* Split TimerTasksFilter into TimerTasksFilter and TimerTasksRangeFilter
* Split SelectFromTimerTasks into SelectFromTimerTasks and RangeSelectFromTimerTasks
* Split DeleteFromTimerTasks into DeleteFromTimerTasks and RangeDeleteFromTimerTasks

**Why?**
This PR provides better test coverage and code management.

**How did you test it?**
Run tests.

**Potential risks**
Minor business logic code change.

### CODE REVIEW

File: common/persistence/sql/sqlplugin/history_transfer_tasks.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		ShardID   int32
+		MinTaskID int64
+		MaxTaskID int64
+	}
+
+	// HistoryTransferTask is the SQL persistence interface for history nodes and history transfer tasks
+	HistoryTransferTask interface {
+		InsertIntoTransferTasks(rows []TransferTasksRow) (sql.Result, error)
+		// SelectFromTransferTasks returns rows that match filter criteria from transfer_tasks table.
+		SelectFromTransferTasks(filter TransferTasksFilter) ([]TransferTasksRow, error)
```
Comment (sergeybykov) [inline common/persistence/sql/sqlplugin/history_transfer_tasks.go:55]: Is this now a point lookup for which both `SharID` and `TaskID` have to be provided? If that's the case, do we even need `TransferTasksFilter` anymore? Should `SelectFromTransferTasks()` take `SharID` and `TaskID` as two arguments instead?

File: common/persistence/sql/sqlplugin/history_transfer_tasks.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		// RangeSelectFromTransferTasks returns rows that match filter criteria from transfer_tasks table.
+		RangeSelectFromTransferTasks(filter TransferTasksRangeFilter) ([]TransferTasksRow, error)
+		// DeleteFromTransferTasks deletes one rows from transfer_tasks table.
+		DeleteFromTransferTasks(filter TransferTasksFilter) (sql.Result, error)
```
Comment (sergeybykov) [inline common/persistence/sql/sqlplugin/history_transfer_tasks.go:59]: Same question about `TransferTasksFilter`.

File: common/persistence/sql/sqlplugin/history_transfer_tasks.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		MaxTaskID int64
+	}
+
+	// HistoryTransferTask is the SQL persistence interface for history nodes and history transfer tasks
+	HistoryTransferTask interface {
+		InsertIntoTransferTasks(rows []TransferTasksRow) (sql.Result, error)
+		// SelectFromTransferTasks returns rows that match filter criteria from transfer_tasks table.
+		SelectFromTransferTasks(filter TransferTasksFilter) ([]TransferTasksRow, error)
+		// RangeSelectFromTransferTasks returns rows that match filter criteria from transfer_tasks table.
+		RangeSelectFromTransferTasks(filter TransferTasksRangeFilter) ([]TransferTasksRow, error)
```
Comment (sergeybykov) [inline common/persistence/sql/sqlplugin/history_transfer_tasks.go:57]: Naming nit: `SelectRange` instead?

File: common/persistence/sql/sqlplugin/history_transfer_tasks.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		// DeleteFromTransferTasks deletes one rows from transfer_tasks table.
+		DeleteFromTransferTasks(filter TransferTasksFilter) (sql.Result, error)
+		// RangeDeleteFromTransferTasks deletes one or more rows from transfer_tasks table.
+		RangeDeleteFromTransferTasks(filter TransferTasksRangeFilter) (sql.Result, error)
```
Comment (sergeybykov) [inline common/persistence/sql/sqlplugin/history_transfer_tasks.go:61]: Same question: `DeleteRange` instead?

File: common/persistence/sql/sqlplugin/history_timer_tasks.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		ShardID             int32
+		VisibilityTimestamp time.Time
+		TaskID              int64
+		Data                []byte
+		DataEncoding        string
+	}
+
+	// TimerTasksFilter contains the column names within timer_tasks table that
+	// can be used to filter results through a WHERE clause
+	TimerTasksFilter struct {
```
Comment (sergeybykov) [inline common/persistence/sql/sqlplugin/history_timer_tasks.go:42]: Are all three fields always required? Same question as with `TransferTasksFilter` do we even need this struct anymore or should we pass the three arguments directly to the methods instead?

File: common/persistence/sql/sqlplugin/history_transfer_tasks.go (hunk 1 above)
Comment (wxing1292) [reply common/persistence/sql/sqlplugin/history_transfer_tasks.go:55]: we can certainly do that.
from my understanding, the filter thing (single parameter vs multiple parameters) is to reduce the amount of work when refactoring / adding variable to the filter.

currently the convention of the SQL side code logic is using a single parameter

File: common/persistence/sql/sqlplugin/history_transfer_tasks.go (hunk 3 above)
Comment (wxing1292) [reply common/persistence/sql/sqlplugin/history_transfer_tasks.go:57]: kind of following the naming convention start from here i guess:
https://github.com/uber/cadence/blob/master/common/persistence/sql/sqlplugin/interfaces.go#L612

@samarabbas @alexshtin what do you think?

File: common/persistence/sql/sqlplugin/history_transfer_tasks.go (hunk 1 above)
Comment (sergeybykov) [reply common/persistence/sql/sqlplugin/history_transfer_tasks.go:55]: If that's the convention we want to follow, that's fine with me. Especially if this is not exposed in the SDK.

File: common/persistence/sql/sqlplugin/history_timer_tasks.go (hunk 5 above)
Comment (wxing1292) [reply common/persistence/sql/sqlplugin/history_timer_tasks.go:42]: all 3 are required for pagination

Comment (wxing1292) [reply common/persistence/sql/sqlplugin/history_timer_tasks.go:42]: and get timer (one and only one)

Comment (sergeybykov) [reply common/persistence/sql/sqlplugin/history_timer_tasks.go:42]: If others are for this convention of using a struct for passing arguments into internal methods, I have no objection.

File: common/persistence/sql/sqlplugin/history_transfer_tasks.go (hunk 1 above)
Comment (wxing1292) [reply common/persistence/sql/sqlplugin/history_transfer_tasks.go:55]: currently the interfaces are purely internal
more need to be done to make the SQL / NoSQL persistence interface public (modularized)

**Review by sergeybykov — APPROVED** [review]

File: common/persistence/sql/sqlplugin/history_timer_tasks.go (hunk 5 above)
Comment (wxing1292) [reply common/persistence/sql/sqlplugin/history_timer_tasks.go:42]: create an issue #810 so we can revisit the naming convention before make the persistence layer modular for override

File: common/persistence/sql/sqlplugin/history_transfer_tasks.go (hunk 3 above)
Comment (alexshtin) [reply common/persistence/sql/sqlplugin/history_transfer_tasks.go:57]: It is not select range, right? It selects transfer tasks but based on some range. I would call it `SelectManyFromTransferTasks`.

Comment (alexshtin) [reply common/persistence/sql/sqlplugin/history_transfer_tasks.go:57]: or `SelectTransferTasksByRange`.

**Pull request merged**